dataset/src/main/scala/frameless/functions/NonAggregateFunctions.scala

```diff
@@ -121,4 +104,156 @@ trait NonAggregateFunctions {
+    new TypedColumn[T, String](untyped.base64(column.untyped))
+  }
+
+  /** Non-Aggregate function: Concatenates multiple input string columns together into a single string column.
+    *
+    * apache/spark
+    */
+  def concat[T](columns: TypedColumn[T, String]*): TypedColumn[T, String] = {
+    new TypedColumn[T, String](untyped.concat(columns.map(_.untyped):_*))
+  }
+
+  /** Non-Aggregate function: Concatenates multiple input string columns together into a single string column,
+    * using the given separator.
+    *
+    * apache/spark
+    */
+  def concat_ws[T](sep: String, columns: TypedColumn[T, String]*): TypedColumn[T, String] = {
+    new TypedColumn[T, String](untyped.concat_ws(sep, columns.map(_.untyped):_*))
+  }
+
+  /** Non-Aggregate function: Locates the position of the first occurrence of substring column
+    * in given string
+    *
+    * @note The position is not zero based, but 1 based index. Returns 0 if substr
+    * could not be found in str.
+    *
+    * apache/spark
+    */
+  def instr[T](column: TypedColumn[T, String], substring: String): TypedColumn[T, Int] = {
+    new TypedColumn[T, Int](untyped.instr(column.untyped, substring))
+  }
+
+  /** Non-Aggregate function: Computes the length of a given string.
+    *
+    * apache/spark
+    */
+  //TODO: Also for binary
+  def length[T](column: TypedColumn[T, String]): TypedColumn[T, Int] = {
+    new TypedColumn[T, Int](untyped.length(column.untyped))
+  }
+
+  /** Non-Aggregate function: Computes the Levenshtein distance of the two given string columns.
+    *
+    * apache/spark
+    */
+  def levenshtein[T](l: TypedColumn[T, String], r: TypedColumn[T, String]): TypedColumn[T, Int] = {
+    new TypedColumn[T, Int](untyped.levenshtein(l.untyped, r.untyped))
+  }
+
+  /** Non-Aggregate function: Converts a string column to lower case.
+    *
+    * apache/spark
+    */
+  def lower[T](e: TypedColumn[T, String]): TypedColumn[T, String] = {
+    new TypedColumn[T, String](untyped.lower(e.untyped))
+  }
+
+  /** Non-Aggregate function: Left-pad the string column with pad to a length of len. If the string column is longer
+    * than len, the return value is shortened to len characters.
+    *
+    * apache/spark
+    */
+  def lpad[T](str: TypedColumn[T, String], len: Int, pad: String): TypedColumn[T, String] = {
+    new TypedColumn[T, String](untyped.lpad(str.untyped, len, pad))
+  }
+
+  /** Non-Aggregate function: Trim the spaces from left end for the specified string value.
+    *
+    * apache/spark
+    */
+  def ltrim[T](str: TypedColumn[T, String]): TypedColumn[T, String] = {
+    new TypedColumn[T, String](untyped.ltrim(str.untyped))
+  }
+
+  /** Non-Aggregate function: Replace all substrings of the specified string value that match regexp with rep.
+    *
+    * apache/spark
+    */
+  def regexp_replace[T](str: TypedColumn[T, String], pattern: Regex, replacement: String): TypedColumn[T, String] = {
```

Comment:
While slightly different from the spark api that takes a `String` rather than `Regex` for the pattern, I believe this is a bit safer as it makes the onus of creating a valid regex pattern more explicit. Were this simply a string, it could non obviously blow up at runtime.

There's also the overload that takes columns at the pattern and replacement https://spark.apache.org/docs/2.2.0/api/scala/index.html#org.apache.spark.sql.functions$@regexp_replace(e:org.apache.spark.sql.Column,pattern:org.apache.spark.sql.Column,replacement:org.apache.spark.sql.Column):org.apache.spark.sql.Column. I'm unsure what the best way about this is. Somehow a `TypedColumn[T, Regex]` with mappings from `Regex` to `String` doesn't feel like it belongs in frameless. Any idea on this or is it best to leave this part of the api out?